File: static/python/queue_with_max.py
from stack_with_max import Stack


# @include
class QueueWithMax:

    def __init__(self):
        self._enqueue = Stack()
        self._dequeue = Stack()

    def enqueue(self, x):
        self._enqueue.push(x)

    def dequeue(self):
        if self._dequeue.empty():
            while not self._enqueue.empty():
                self._dequeue.push(self._enqueue.pop())
        if not self._dequeue.empty():
            return self._dequeue.pop()
        raise IndexError('empty queue')

    def max(self):
        if not self._enqueue.empty():
            return self._enqueue.max() if self._dequeue.empty() else max(
                self._enqueue.max(), self._dequeue.max())
        if not self._dequeue.empty():
            return self._dequeue.max()
        raise IndexError('empty queue')
# @exclude


def simple_test():
    Q = QueueWithMax()
    Q.enqueue(11)
    Q.enqueue(2)
    assert 11 == Q.max()
    assert 11 == Q.dequeue()
    assert 2 == Q.max()
    assert 2 == Q.dequeue()
    Q.enqueue(3)
    assert 3 == Q.max()
    assert 3 == Q.dequeue()
    maxint = 2**64 - 1
    Q.enqueue(maxint - 1)
    Q.enqueue(maxint)
    Q.enqueue(-2)
    Q.enqueue(-1)
    Q.enqueue(-1)
    minint = -2**64
    Q.enqueue(minint)
    assert maxint == Q.max()
    assert maxint - 1 == Q.dequeue()
    assert maxint == Q.max()
    assert maxint == Q.dequeue()
    assert -1 == Q.max()
    assert -2 == Q.dequeue()
    assert -1 == Q.max()
    assert -1 == Q.dequeue()
    assert -1 == Q.dequeue()
    assert minint == Q.max()
    assert minint == Q.dequeue()
    try:
        print('Q is empty, max() call should except = ' % Q.max())
        assert False
    except IndexError as e:
        print(e)  # throw


def main():
    simple_test()
    Q = QueueWithMax()
    Q.enqueue(1)
    Q.enqueue(2)
    assert 2 == Q.max()
    assert 1 == Q.dequeue()  # 1
    assert 2 == Q.max()
    assert 2 == Q.dequeue()  # 2
    Q.enqueue(3)
    assert 3 == Q.max()
    assert 3 == Q.dequeue()  # 3
    try:
        Q.max()
        assert False
    except IndexError as e:
        print(e)  # throw
    try:
        Q.dequeue()
        assert False
    except IndexError as e:
        print(e)  # throw


if __name__ == '__main__':
    main()

File: static/solutions/Queue_with_max.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <limits>
#include <stdexcept>

#include "./Stack_with_max.h"

using std::cout;
using std::endl;
using std::exception;
using std::length_error;
using std::max;
using std::numeric_limits;

// @include
class QueueWithMax {
 public:
  void Enqueue(int x) { enqueue_.Push(x); }

  int Dequeue() {
    if (dequeue_.Empty()) {
      while (!enqueue_.Empty()) {
        dequeue_.Push(enqueue_.Pop());
      }
    }
    if (!dequeue_.Empty()) {
      return dequeue_.Pop();
    }
    throw length_error("empty queue");
  }

  int Max() const {
    if (!enqueue_.Empty()) {
      return dequeue_.Empty() ? enqueue_.Max()
                              : max(enqueue_.Max(), dequeue_.Max());
    }
    if (!dequeue_.Empty()) {
      return dequeue_.Max();
    }
    throw length_error("empty queue");
  }

 private:
  Stack enqueue_, dequeue_;
};
// @exclude

void SimpleTest() {
  QueueWithMax Q;
  Q.Enqueue(11);
  Q.Enqueue(2);
  assert(11 == Q.Max());
  assert(11 == Q.Dequeue());
  assert(2 == Q.Max());
  assert(2 == Q.Dequeue());
  Q.Enqueue(3);
  assert(3 == Q.Max());
  assert(3 == Q.Dequeue());
  Q.Enqueue(numeric_limits<int>::max() - 1);
  Q.Enqueue(numeric_limits<int>::max());
  Q.Enqueue(-2);
  Q.Enqueue(-1);
  Q.Enqueue(-1);
  Q.Enqueue(numeric_limits<int>::min());
  assert(numeric_limits<int>::max() == Q.Max());
  assert(numeric_limits<int>::max() - 1 == Q.Dequeue());
  assert(numeric_limits<int>::max() == Q.Max());
  assert(numeric_limits<int>::max() == Q.Dequeue());
  assert(-1 == Q.Max());
  assert(-2 == Q.Dequeue());
  assert(-1 == Q.Max());
  assert(-1 == Q.Dequeue());
  assert(-1 == Q.Dequeue());
  assert(numeric_limits<int>::min() == Q.Max());
  assert(numeric_limits<int>::min() == Q.Dequeue());
  try {
    cout << "Q is empty, Max() call should except = " << Q.Max();
    assert(false);
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
}

int main(int argc, char* argv[]) {
  SimpleTest();
  QueueWithMax Q;
  Q.Enqueue(1);
  Q.Enqueue(2);
  assert(2 == Q.Max());
  assert(1 == Q.Dequeue());
  assert(2 == Q.Max());
  assert(2 == Q.Dequeue());
  Q.Enqueue(3);
  assert(3 == Q.Max());
  assert(3 == Q.Dequeue());
  try {
    Q.Max();
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
  try {
    Q.Dequeue();
  } catch (const exception& e) {
    cout << e.what() << endl;
  }
  return 0;
}
